### PR TITLE
Fix stuck in local minimum / excess loops in path

### PR DESCRIPTION
**Pull request by howardcochran**
These two commits provide a practical solution to the problem of the
planner getting stuck in a sub-optimal local minimum. Commonly, these
local minima cause the robot to perform excess loops on its way to the
goal. The solution solves the problem probabilistically by allowing the
multi-homotopy class planner to maintain a second candidate path in the
same homotopy class as the current "best" path. This allows path
exploration to "discover" new paths in a different, potentially better
local minimum, typically leading to the most optimal such alternative
being found.

This solves the following open issues:
https://github.com/rst-tu-dortmund/teb_local_planner/issues/88
https://github.com/rst-tu-dortmund/teb_local_planner/issues/95
https://github.com/rst-tu-dortmund/teb_local_planner/issues/115
https://github.com/rst-tu-dortmund/teb_local_planner/issues/135

It may also address this issue, but I am uncertain what the author means
by "teb cannot converge the path to a good solution.":
https://github.com/rst-tu-dortmund/teb_local_planner/issues/61

The second commit in this series not only improves the chances of
getting out of a local minimum but also improves the chances of finding
a better path around obstacles because it does not keep all the
candidate slots full with only the first few homotopy classes discovered
in exploration.

### CODE REVIEW

**Comment by croesmann** [comment]
Thanks a lot, looks good to me and I was able to test this successfully. 
However, I have two minor points:
- I would like to set `selection_pruning_probability` to zero for dynamic reconfiguration in order to retain the old behavior, especially in kinetic. We can think of changing this in future distribution. However, I am going to include a value of 0.3 in the teb_local_planner_tutorials configuration files.
- What do you think of renaming `pruneTebs()` to `dropTebs()` in order to avoid confusions with the meaning of pruning in `timed_elastic_band.cpp`?

**Comment by howardcochran** [comment]
> Thanks a lot, looks good to me and I was able to test this successfully.

Yay, glad to hear it.

> However, I have two minor points:
> 
> * I would like to set `selection_pruning_probability` to zero for dynamic reconfiguration in order to retain the old behavior, especially in kinetic. We can think of changing this in future distribution. However, I am going to include a value of 0.3 in the teb_local_planner_tutorials configuration files.

Sure.  I set the default to 0.0 and renamed the parameter `selection_dropping_probability`

> * What do you think of renaming `pruneTebs()` to `dropTebs()` in order to avoid confusions with the meaning of pruning in `timed_elastic_band.cpp`?

I renamed `pruneTebs()` to `randomlyDropTebs()`. I also made it return immediately if `selection_dropping_probability` is zero.

And, of course, I rebased onto current kinetic-devel.
Thank you!

**Comment by howardcochran** [comment]
Since my original submission, upstream changes will break this.  I will fix this for compatibility, rebase, and resubmit shortly.

**Comment by howardcochran** [comment]
OK. I have rebased and force-pushed these commits to fix a compatibility issue with
43ea03209 ("Never discarding the previous best teb in renewAndAnalyzeOldTebs")

Should be ready for merging.

**Comment by croesmann** [comment]
thank you @howardcochran ! I'll test this again soon!

**Comment by howardcochran** [comment]
> thank you @howardcochran ! I'll test this again soon!

Thanks. Let me know if you want additional tweaks.

**Comment by AntonBock** [comment]
I'm having a problem that sounds like what you have fixed @howardcochran. As you can see in the picture, my robot gets stuck trying to go through a wall, instead of going around. Is there anyway to download the fix, or should I just wait until it gets merged?

LocalMinimum

**Comment by croesmann** [comment]
@AntonBock Could you try to clone this merge request and tell me if it fixes your issue?

**Comment by AntonBock** [comment]
@croesmann I have cloned the merge request into the melodic devel (had to add an #include <random> in a header file). Unfortunately, it has not fixed the issue. The planner still tries to "teleport" through the wall instead of going around it. Any idea what could cause this? Should I make a new issue?

image

**Comment by maxlein** [comment]
Are you sure that teb planner knows about your local obstacles?

**Comment by AntonBock** [comment]
@maxlein It seems like the planner sees the obstacles. If the obstacle is smaller, ie. just a small traffic cone, the planner will easily make a path around it. I don't know if you can see it in the picture, but the planner doesn't seem to be placing the teb_markers through the wall, although the local path is planned there. This looks like the same problem posed in the ros tutorial found here: https://wiki.ros.org/teb_local_planner/Tutorials/Obstacle%20Avoidance%20and%20Robot%20Footprint%20Model, but no solution is given.

**Comment by howardcochran** [comment]
> @croesmann I have cloned the merge request into the melodic devel (had to add an #include in a header file). Unfortunately, it has not fixed the issue. The planner still tries to "teleport" through the wall instead of going around it. Any idea what could cause this? Should I make a new issue?

@AntonBock, The issue that you illustrate appears very different from what my fix addresses (more on that below).
From the images you have posted, it appears that teb was initialized with an invalid plan (one that goes through the wall), and that it failed to find a valid one. For the most part, Teb depends on being initialized with a global plan that is at least topologically valid. In other words, it works best if the plan provided by the global planner goes around obstacles. Teb will then optimze the global plan from something which may be very sub-optimal and even kinesmatically infeasible into one which is kinesmatically feasible and which optimizes various tunable critera.

If given a "straight line" plan through a wall, Teb will generally push the poses forward and backward away from the wall, but there is nothing in the optimizer graph to push the plan around the wall.

HOWEVER, it may be able to find a way around the wall when you enable its Homotopy Class Planning (by setting the enable_homotopy_class_planning to true and max_number_classes parameter >= 2). This will cause it to explore random paths, which may result in it finding one that goes around that wall. But this can take quite some time and is not guaranteed to happen.

Bottom Line: Look at the configuration of your global planner and make sure it is producing a plan that goes around the wall and that this is the plan that teb is receiving.

About my change: It helps keeps Teb from getting stuck with a sub-optimal (but valid) plan  containing one or more extra loops. In your case, however, it is not finding a valid plan to begin with, so my change is not able to help.

**Comment by AntonBock** [comment]
@howardcochran thank you for the detailed response! I'll look into the global planner, and see if that fixes my issue. Sorry for the misunderstanding

**Comment by howardcochran** [comment]
> @howardcochran thank you for the detailed response! I'll look into the global planner, and see if that fixes my issue. Sorry for the misunderstanding

@AntonBock, you are welcome. Best of luck!

**Comment by ljburtz** [comment]
has this been merged? looks like it's exactly the fix I need for the issue of the planner producing all these loops! Thanks for this awesomeness!

**Comment by amakarow** [comment]
@ros-pull-request-builder retest this please

**Comment by amakarow** [comment]
I just rebased this PR to melodic and merged it. Many thanks for your contribution.
